Sony Touchpad Fix v2

TVsIan opened the pull request:
Replaces https://github.com/batocera-linux/batocera.linux/pull/7182

New method: uses evdev to determine what is a mouse or gamepad based on available inputs. If a mouse and gamepad have the same physical ID (checks out on BT and USB), the input device of the controller is stored. If there's a mouse with no matching gamepad, a dummy value gets stored to keep proper indexing.

When Retroarch generates the controller config, if the system is NDS or 3DS (we assume other systems will want a real mouse, but others can be added), it will check to see if the current controller matches one of the input devices in the list. If so, it will set the mouse index to the index in the list.

dmanlfc commented:
@nadenislamarre - looks better

TVsIan commented:
Slight modification - testing looks like Retroarch's mouse indexes are ordered by the device's event ID. The sorting has been changed to adjust for that.

nadenislamarre commented:
yes, but i need to analyze to understand why we do that. sorry, i'm quite buzy these time

nadenislamarre commented:
if 'BTN_LEFT' in keyString or 'BTN_TOUCH' in keyString:
                mouseDevices[checkInput.phys] = deviceList
            if 'BTN_SELECT' in keyString or 'BTN_START' in keyString or 'BTN_SOUTH' in keyString:
                gamepadDevices[checkInput.phys] = deviceList


this is not nice.
a mouse is not device with some specific buttons.
a mouse is a device with the mouse flag.
https://github.com/batocera-linux/batocera.linux/blob/master/package/batocera/core/batocera-configgen/configgen/configgen/controllersConfig.py#L267

- if controllers[controller].dev in mouseList:
i don't understand what we check here.
the "Sony Touchpad" : which pad is it ?
it is mouse and joystick at the same time ? (possible)
or it is on 2 different devices ?

if the joystick is a mouse at the same time, i guess we're near the correct implementation.
i would rename generateMouseList to getGuns and put it next to it to be similar, and just check the mouse flag.

TVsIan commented:
The reason for this function is that Retroarch will assign mouse index 0 to
the player 1 mouse input. In the case of having a DS4 or DualSense and a
mouse connected, index 0 is most likely the mouse and not the touchpad on
the controller. For at least NDS/3DS, using the trackpad is preferable to a
mouse.

At some point in the last, Retroarch used device names for the mouse. The
indexes can change on reboot or if devices are swapped around, so the
current method can’t guarantee the same device every time.

After some testing, it looks like Retroarch’s mouse indexes are based on
the event ID # - in order, mouse devices only. So we build a list of mouse
devices in that order, then see if the physical ID of any of those mice
match a controller. We make a list of the controller’s event file location
in the same order, with a dummy item for mouse-only devices to keep the
indexes matching. When Retroarch is setting up a controller for NDS/3DS, if
the event ID of the controller is found in the list, it applies that index
to the mouse index, so if the port is using a Sony controller, the mouse
device will be that controller’s trackpad.

We don’t do this for other systems because for computers etc a mouse would
be preferable. I can modify the code to match trackpads if the only mouse
devices match controllers if it’s needed.

I have plans on expanding this functionality to properly assign trackball,
spinner, etc. for arcade games in Retroarch (requested in Discord
#suggestions), this function is just to fix an oversight in Retroarch’s
mouse assignment.

On Fri, Oct 21, 2022 at 11:27 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> yes, but i need to analyze to understand why we do that. sorry, i'm quite
> buzy these time
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1287121893>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2W6ODF5EZTROU3MLRTWEKY7JANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
The version of Python evdev we’re currently using seems to be missing the
property function. The buttons were the only available method I could find
to determine the device type.

If we could update that module, I can change it.

And basically a Sony controller creates three events: controller, gyro, and
touchpad. The touchpad is a mouse, I haven’t checked to see what the gyro
device comes up as. This matches the mouse and gamepad devices.

On Fri, Oct 21, 2022 at 11:37 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

>         if 'BTN_LEFT' in keyString or 'BTN_TOUCH' in keyString:
>             mouseDevices[checkInput.phys] = deviceList
>         if 'BTN_SELECT' in keyString or 'BTN_START' in keyString or 'BTN_SOUTH' in keyString:
>             gamepadDevices[checkInput.phys] = deviceList
>
> this is not nice.
> a mouse is not device with some specific buttons.
> a mouse is a device with the mouse flag.
>
> https://github.com/batocera-linux/batocera.linux/blob/master/package/batocera/core/batocera-configgen/configgen/configgen/controllersConfig.py#L267
>
>    - if controllers[controller].dev in mouseList:
>    i don't understand what we check here.
>    the "Sony Touchpad" : which pad is it ?
>    it is mouse and joystick at the same time ? (possible)
>    or it is on 2 different devices ?
>
> if the joystick is a mouse at the same time, i guess we're near the
> correct implementation.
> i would rename generateMouseList to getGuns and put it next to it to be
> similar, and just check the mouse flag.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1287133773>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2XS2RH6MBZJN2D5RFTWEK2EHANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nadenislamarre commented:
how do you do the link between the touchpad device and the joystick device ?

TVsIan commented:
That function matches the physical location of a mouse and gamepad device.
Testing with multiple controllers connected shows that matches (between all
3 events) every time.

Since our controller type includes the event ID, the function spits out a
list where the items/order match the mouse devices, and the contents are
the event of the matching controller.

So with one mouse and one Sony controller, the list may be something like [
“dummy”, “/dev/input/event8” ] - the dummy entry is to make sure the
indexes are maintained.

In the Retroarch controller config generator, it checks the list, and if
the controller currently being added is in that list, it sets Retroarch’s
mouse index for that controller to the index of that item in the list. So
in this case, the controller that has event8 gets mouseindex 1 in Retroarch.

Without this function, port 1 would have gotten mouse index 0. So port 1
would use the actual mouse. And if there were a player 2, their mouse would
be set to the trackpad on player 1’s controller.

On Fri, Oct 21, 2022 at 11:50 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> how do you do the link between the touchpad device and the joystick device
> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1287147049>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2V736SDOT6GV6FCKDDWEK3SRANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nadenislamarre commented:
sorry, i don't understand the aim of this pr.
i don't understand the initial issue in fact.
a ds4 is a ps4 controller ?

TVsIan commented:
Yes, the DualShock 4. The issue is that Retroarch is not assigning the
controller’s touchpad (which is a separate mouse device) to the input port
that matches the controller.

On Mon, Oct 24, 2022 at 2:16 PM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> sorry, i don't understand the aim of this pr.
> i don't understand the initial issue in fact.
> a ds4 is a ps4 controller ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1289411624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2TUJFSL6BA5CQWQFEDWE3HAZANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
Sure, that function matches mouse and gamepad devices based on them sharing
a physical ID.

The output list corresponds to the Retroarch mouse indexes, and the
contents are the id of the matching controller. So indexes 0 & 1 are dummy
just to hold their place in the list, they have no matching controller.
Index 2 is the controller that matches that mouse device.

The use case is Nintendo DS/3DS, the preferred behavior is to use the
touchpad on the controller to control the touchscreen. This will only apply
to those two systems so as not to interfere with lightguns or mice.

The Retroarch controller config will check the event of the controller
that’s being configured, and if it is in the list, will set the mouseindex
to the index of the item in the list (I could swap the list for a
dictionary if that would be cleaner).

Retroarch does not save mouse indexes, and even if they did, they can
change on reboot. So the user would have to go into the Retroarch settings
every launch to use the touchpad.

On Mon, Oct 24, 2022 at 2:42 PM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> ok.
> what is the usecase ?
>
> now i try to understand the idea behind generateMousesList()
> i don't understand it, so, i execute to help me.
> i get :
> ['dummy', 'dummy', '/dev/input/event17']
>
> so, i read on my board what looks like a mouse, and i have event6 and
> event15
> event17 is joystick.
> i don't know if the function name is wrong or if i miss something.
> can you light me please ?
>
> ***@***.*** /userdata/system/tmp]# for i in /dev/input/event*; do
> udevadm info -q all -n $i | grep ID_INPUT_MOUSE && echo $i; done
> E: ID_INPUT_MOUSE=1
> /dev/input/event6
> ***@***.*** /userdata/system/tmp]# for i in /dev/input/event*; do
> udevadm info -q all -n $i | grep ID_INPUT_TOUCHPAD && echo $i; done
> E: ID_INPUT_TOUCHPAD=1
> /dev/input/event15
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1289444501>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2RMRG5ONP2SFD23RILWE3J7TANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nadenislamarre commented:
i think i get what's you're trying to do
device /dev/input/event17, name "Sony Interactive Entertainment Wireless Controller", phys "usb-0000:04:00.3-1.1/input3"
/dev/input/event16
device /dev/input/event16, name "Sony Interactive Entertainment Wireless Controller Motion Sensors", phys "usb-0000:04:00.3-1.1/input3"
/dev/input/event15
device /dev/input/event15, name "Sony Interactive Entertainment Wireless Controller Touchpad", phys "usb-0000:04:00.3-1.1/input3"

TVsIan commented:
It also looks like you have a third device that is showing mouse buttons in
its input, but not showing as a mouse device. Possibly a lightgun? It would
be worth comparing that to the list of mice in Retroarch to see if there
are 2 or 3 devices in the list.

On Mon, Oct 24, 2022 at 2:51 PM Ian Murray ***@***.***> wrote:

> Sure, that function matches mouse and gamepad devices based on them
> sharing a physical ID.
>
> The output list corresponds to the Retroarch mouse indexes, and the
> contents are the id of the matching controller. So indexes 0 & 1 are dummy
> just to hold their place in the list, they have no matching controller.
> Index 2 is the controller that matches that mouse device.
>
> The use case is Nintendo DS/3DS, the preferred behavior is to use the
> touchpad on the controller to control the touchscreen. This will only apply
> to those two systems so as not to interfere with lightguns or mice.
>
> The Retroarch controller config will check the event of the controller
> that’s being configured, and if it is in the list, will set the mouseindex
> to the index of the item in the list (I could swap the list for a
> dictionary if that would be cleaner).
>
> Retroarch does not save mouse indexes, and even if they did, they can
> change on reboot. So the user would have to go into the Retroarch settings
> every launch to use the touchpad.
>
> On Mon, Oct 24, 2022 at 2:42 PM Nicolas Adenis-Lamarre <
> ***@***.***> wrote:
>
>> ok.
>> what is the usecase ?
>>
>> now i try to understand the idea behind generateMousesList()
>> i don't understand it, so, i execute to help me.
>> i get :
>> ['dummy', 'dummy', '/dev/input/event17']
>>
>> so, i read on my board what looks like a mouse, and i have event6 and
>> event15
>> event17 is joystick.
>> i don't know if the function name is wrong or if i miss something.
>> can you light me please ?
>>
>> ***@***.*** /userdata/system/tmp]# for i in /dev/input/event*; do
>> udevadm info -q all -n $i | grep ID_INPUT_MOUSE && echo $i; done
>> E: ID_INPUT_MOUSE=1
>> /dev/input/event6
>> ***@***.*** /userdata/system/tmp]# for i in /dev/input/event*; do
>> udevadm info -q all -n $i | grep ID_INPUT_TOUCHPAD && echo $i; done
>> E: ID_INPUT_TOUCHPAD=1
>> /dev/input/event15
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1289444501>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AE5KD2RMRG5ONP2SFD23RILWE3J7TANCNFSM6AAAAAARAOQAP4>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

TVsIan commented:
Yes, that’s the issue that we’re trying to work around, Retroarch can’t
match the different devices together.

On Mon, Oct 24, 2022 at 2:53 PM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> i think i get what's you're trying to do
> device /dev/input/event17, name "Sony Interactive Entertainment Wireless
> Controller", phys "usb-0000:04:00.3-1.1/input3"
> /dev/input/event16
> device /dev/input/event16, name "Sony Interactive Entertainment Wireless
> Controller Motion Sensors", phys "usb-0000:04:00.3-1.1/input3"
> /dev/input/event15
> device /dev/input/event15, name "Sony Interactive Entertainment Wireless
> Controller Touchpad", phys "usb-0000:04:00.3-1.1/input3"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1289456687>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2UQRJIJLDVYOY7PM6LWE3LJPANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nadenislamarre commented:
yes, i've patched retroarch so that at least the mouse number is the mouse with the lowest event number. before, it was more or less random.

nadenislamarre commented:
by the way, your function is slow. probably due to the lib or the call.
let me write a function that may be helpful.

nadenislamarre commented:
i've done this function :

https://pastebin.com/raw/7X5HtZqW

it is very quick.
it is reusable (maybe is should include guns).
but maybe it is "too" complex.

the result on my side :

{'/dev/input/event14': {'associatedDevices': [],
                        'isJoystick': True,
                        'isMouse': False,
                        'joystick_index': 0,
                        'mouse_index': None},
 '/dev/input/event15': {'associatedDevices': ['/dev/input/event17'],
                        'isJoystick': False,
                        'isMouse': True,
                        'joystick_index': None,
                        'mouse_index': 1},
 '/dev/input/event17': {'associatedDevices': ['/dev/input/event15'],
                        'isJoystick': True,
                        'isMouse': False,
                        'joystick_index': 1,
                        'mouse_index': None},
 '/dev/input/event6': {'associatedDevices': [],
                       'isJoystick': False,
                       'isMouse': True,
                       'joystick_index': None,
                       'mouse_index': 0}}

in your case,
from /dev/input/event17 you jump to /dev/input/event15 and find it is a mouse.
note that i patched all needed emulators so that indexes are the same.
it is required for simpler configurations ; if not mandatory when some emulators write in their configuration a pad number which is more or less random...

TVsIan commented:
That looks like it would work for what this needs, I can just generate a
dictionary using associateddevices: mouseindex and pass that to the
libretro controls. Once I finish what I'm working on, I'll modify this PR
to use that function.

On Mon, Oct 24, 2022 at 4:20 PM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> i've done this function :
>
> https://pastebin.com/raw/7X5HtZqW
>
> it is very quick.
> it is reusable (maybe is should include guns).
> but maybe it is "too" complex.
>
> the result on my side :
>
> {'/dev/input/event14': {'associatedDevices': [],
> 'isJoystick': True,
> 'isMouse': False,
> 'joystick_index': 0,
> 'mouse_index': None},
> '/dev/input/event15': {'associatedDevices': ['/dev/input/event17'],
> 'isJoystick': False,
> 'isMouse': True,
> 'joystick_index': None,
> 'mouse_index': 1},
> '/dev/input/event17': {'associatedDevices': ['/dev/input/event15'],
> 'isJoystick': True,
> 'isMouse': False,
> 'joystick_index': 1,
> 'mouse_index': None},
> '/dev/input/event6': {'associatedDevices': [],
> 'isJoystick': False,
> 'isMouse': True,
> 'joystick_index': None,
> 'mouse_index': 0}}
>
> in your case,
> from /dev/input/event17 you jump to /dev/input/event15 and find it is a
> mouse.
> note that i patched all needed emulators so that indexes are the same.
> it is required for simpler configurations ; if not mandatory when some
> emulators write in their configuration a pad number which is more or less
> random...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7227#issuecomment-1289565033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2QMOSJ3VRVMN4P6JS3WE3VSHANCNFSM6AAAAAARAOQAP4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nadenislamarre commented:
you can add a such function :
https://pastebin.com/raw/5pT58exv

TVsIan commented:
I replaced my function with your two functions, and rewrote the modified part of the libretro controller config to use them instead. Tested, and it appears to be working as expected.

nadenislamarre commented:
import pprint must be removed. it was just to display the example.
for the rest, it is ok.

TVsIan commented:
Done.
